zebra-state/src/lib.rs

```diff
@@ -142,10 +155,27 @@ pub enum Response {
 
 /// Get the heights of the blocks for constructing a block_locator list
 fn block_locator_heights(tip_height: BlockHeight) -> impl Iterator<Item = BlockHeight> {
-    iter::successors(Some(1u32), |h| h.checked_mul(2))
-        .flat_map(move |step| tip_height.0.checked_sub(step))
+    // Stop at the reorg limit, or the genesis block.
+    let min_locator_height = tip_height
+        .0
+        .checked_sub(MAX_BLOCK_REORG_HEIGHT.0)
         .map(BlockHeight)
-        .chain(iter::once(BlockHeight(0)))
+        .unwrap_or(BlockHeight(0));
+    let locators = iter::successors(Some(1u32), |h| h.checked_mul(2))
+        .flat_map(move |step| tip_height.0.checked_sub(step))
+        .filter(move |&height| height > min_locator_height.0)
+        .map(BlockHeight);
+    let locators = iter::once(tip_height)
+        .chain(locators)
+        .chain(iter::once(min_locator_height));
```

Comment:
I think this could be reduced to a single chain call and a `take_while`, instead of the two chains and the filter.

Reply:
I'm not sure exactly how to do that, but I did find some things that could be simplified.

I'm going to write some tests, because I also discovered another subtle bug, where we would repeat the genesis hash twice.

Reply:
I managed to simplify the code a lot in c10a61b43f2fe2534aa3ef58049130566a8b1329, but there are still 3 chain calls.

Let me know if you can find any more simplifications. I've written the tests so they enforce basic properties like min height, max height, and uniqueness. But the interim heights are flexible.